File: js/applications/hotPotato.ts
import { Queue } from "@lib/queue";

export function hotPotato(elementsList: string[], num: number) {
    const queue = new Queue<string>();
    const eliminatedList: string[] = [];

    for (let i = 0; i < elementsList.length; i++) {
        queue.enqueue(elementsList[i]);
    }

    while (queue.size() > 1) {
        for (let i = 0; i < num; i++) {
            // @ts-ignore
            queue.enqueue(queue.dequeue());
        }
        // @ts-ignore
        eliminatedList.push(queue.dequeue());
    }

    return {
        eliminated: eliminatedList,
        winner: queue.dequeue()
    };
}


const names = ['John', 'Doe', 'Maria', 'Joao']
const response = hotPotato(names, 2)

console.table(response)

File: lua/applications/hot_potato.lua
require('lib/queue')

function HotPotato(list, num)
    local queue = Queue.new()
    local eliminated = Queue.new()

    for i = 1, #list, 1 do
        queue:enqueue(list[i])
    end

    while queue:size() > 1 do
        for _ = 1, num, 1 do
            queue:enqueue(queue:dequeue())
        end

        eliminated:enqueue(queue:dequeue())
    end

    return eliminated, queue:dequeue()
end

local names = {
    "John", "Jack", "Camila", "Ingrid", "Carl"
}

local eliminated, winner = HotPotato(names, 7)

for i = 1, eliminated:size(), 1 do
    print("eliminated: ", eliminated.items[i])
end

print("winner: ", winner)

